File: Samples/C#/ModelSelection/ModelSelectionPlugin.cs
/* Sample: Model Selection
 * 
 * This sample illustrates how work with the selection mechanism of Renga API.
 * 
 * Copyright Renga Software LLC, 2018. All rights reserved.
 */

using System;
using System.Collections.Generic;
using System.Text;

namespace ModelSelection
{
  public class ModelSelectionPlugin : Renga.IPlugin
  {
    private Renga.IApplication m_application;

    private Renga.IAction m_showSelectedObjectsAction;
    private Renga.ActionEventSource m_showSelectedObjectsActionEventSource;
    private Renga.ActionEventSource m_selectWallsActionEventSource;

    private Renga.SelectionEventSource m_selectionEventSource;

    public bool Initialize(string pluginFolder)
    {
      m_application = new Renga.Application();

      var selection = m_application.Selection;
      m_selectionEventSource = new Renga.SelectionEventSource(selection);
      m_selectionEventSource.ModelSelectionChanged += OnModelSelectionChanged;

      var ui = m_application.UI;

      var uiPanelExtension = ui.CreateUIPanelExtension();

      m_showSelectedObjectsAction = ui.CreateAction();
      m_showSelectedObjectsAction.ToolTip = "Show selected objects";
      m_showSelectedObjectsAction.Enabled = selection.GetSelectedObjects().Length > 0;
      m_showSelectedObjectsActionEventSource = new Renga.ActionEventSource(m_showSelectedObjectsAction);
      m_showSelectedObjectsActionEventSource.Triggered += OnShowSelectedObjectsButtonClicked;
      uiPanelExtension.AddToolButton(m_showSelectedObjectsAction);

      var selectWallsAction = ui.CreateAction();
      selectWallsAction.ToolTip = "Select walls";
      m_selectWallsActionEventSource = new Renga.ActionEventSource(selectWallsAction);
      m_selectWallsActionEventSource.Triggered += OnSelectWallsButtonClicked;
      uiPanelExtension.AddToolButton(selectWallsAction);

      ui.AddExtensionToPrimaryPanel(uiPanelExtension);

      return true;
    }

    public void Stop()
    {
      m_selectionEventSource.Dispose();
      m_showSelectedObjectsActionEventSource.Dispose();
      m_selectWallsActionEventSource.Dispose();
    }

    private void OnModelSelectionChanged(object sender, EventArgs args)
    {
      var selection = m_application.Selection;
      m_showSelectedObjectsAction.Enabled = selection.GetSelectedObjects().Length > 0;
    }

    private void OnShowSelectedObjectsButtonClicked(object sender, EventArgs args)
    {
      var project = m_application.Project;

      if (project == null)
        return;

      var model = project.Model;
      var modelObjectCollection = model.GetObjects();

      var selection = m_application.Selection;

      int[] selectedObjectIds = (int[])selection.GetSelectedObjects();

      var stringBuilder = new StringBuilder();
      stringBuilder.AppendLine($"{selectedObjectIds.Length} selected objects:");

      foreach (int objectId in selectedObjectIds)
      {
        stringBuilder.AppendLine($"Object id: {objectId}");
      }

      var ui = m_application.UI;
      ui.ShowMessageBox(Renga.MessageIcon.MessageIcon_Info, "ModelSelection", stringBuilder.ToString());
    }

    private void OnSelectWallsButtonClicked(object sender, EventArgs args)
    {
      var project = m_application.Project;

      if (project == null)
        return;

      var model = project.Model;
      var modelObjects = model.GetObjects();

      int[] modelObjectIds = (int[])modelObjects.GetIds();

      var wallObjectIds = new List<int>();

      foreach (int objectId in modelObjectIds)
      {
        var modelObject = modelObjects.GetById(objectId);

        if (modelObject == null)
          continue;

        if (modelObject.ObjectType != Renga.ObjectTypes.Wall)
          continue;

        wallObjectIds.Add(objectId);
      }

      var selection = m_application.Selection;
      selection.SetSelectedObjects(wallObjectIds.ToArray());
    }
  }
}

File: Samples/C++/ModelSelection/ModelSelectionPlugin.cpp
/*  Sample: Model Objects
 *
 *  This sample illustrates how to traverse model objects in a project and read their parameters.
 *
 *  Copyright Renga Software LLC, 2018. All rights reserved.
 */

#include "ModelSelectionPlugin.h"

#include <sstream>
#include <vector>

#include <Renga/ObjectTypes.h>
#include <Renga/CreateApplication.hpp>

#include <atlsafe.h>

ModelSelectionPlugin::ModelSelectionPlugin()
{
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);
}

ModelSelectionPlugin::~ModelSelectionPlugin()
{
  ::CoUninitialize();
}

bool ModelSelectionPlugin::initialize(const wchar_t* pluginDir)
{
  auto pApplication = Renga::CreateApplication();
  if (!pApplication)
    return false;

  auto pSelection = pApplication->GetSelection();
  if (!pSelection)
    return false;

  if (auto pUI = pApplication->GetUI())
  {
    if (auto pUIPanelExtension = pUI->CreateUIPanelExtension())
    {
      if (auto pShowSelectedObjectsAction = pUI->CreateAction())
      {
        m_pModelSelectionChangedHandler.reset(new ModelSelectionChangedHandler(pSelection, pShowSelectedObjectsAction));
        
        pShowSelectedObjectsAction->PutToolTip(L"Show selected objects");
        m_pShowSelectedObjectsButtonClickHandler.reset(new ShowSelectedObjectsButtonClickHandler(pShowSelectedObjectsAction, pApplication));
        pUIPanelExtension->AddToolButton(pShowSelectedObjectsAction);
      }

      if (auto pSelectWallsAction = pUI->CreateAction())
      {
        pSelectWallsAction->PutToolTip(L"Select walls");
        m_pSelectWallsButtonClickHandler.reset(new SelectWallsButtonClickHandler(pSelectWallsAction, pApplication));
        pUIPanelExtension->AddToolButton(pSelectWallsAction);
      }

      pUI->AddExtensionToPrimaryPanel(pUIPanelExtension);
    }
  }

  return true;
}

void ModelSelectionPlugin::stop()
{
  m_pShowSelectedObjectsButtonClickHandler.reset();
  m_pSelectWallsButtonClickHandler.reset();
}

ModelSelectionChangedHandler::ModelSelectionChangedHandler(Renga::ISelectionPtr pSelection, Renga::IActionPtr showSelectedObjectsAction)
  : Renga::SelectionEventHandler(pSelection),
    m_pSelection(pSelection),
    m_pShowSelectedObjectsAction(showSelectedObjectsAction)
{
  UpdateShowSelectedObjectsButton();
}

void ModelSelectionChangedHandler::OnModelSelectionChanged()
{
  UpdateShowSelectedObjectsButton();
}

void ModelSelectionChangedHandler::UpdateShowSelectedObjectsButton()
{
  CComSafeArray<int> selectedObjectIds(m_pSelection->GetSelectedObjects());
  m_pShowSelectedObjectsAction->PutEnabled(selectedObjectIds.GetCount() > 0);
}

ShowSelectedObjectsButtonClickHandler::ShowSelectedObjectsButtonClickHandler(Renga::IActionPtr pAction, Renga::IApplicationPtr pApplication)
  : Renga::ActionEventHandler(pAction),
    m_pApplication(pApplication)
{
}

void ShowSelectedObjectsButtonClickHandler::OnTriggered()
{
  auto pProject = m_pApplication->GetProject();

  if (pProject == nullptr)
    return;

  auto pModel = pProject->GetModel();
  auto pModelObjectCollection = pModel->GetObjects();

  auto pSelection = m_pApplication->GetSelection();

  CComSafeArray<int> selectedObjectIds(pSelection->GetSelectedObjects());

  std::wstringstream stringStream;
  stringStream << selectedObjectIds.GetCount() << L" selected objects:" << std::endl;

  for (int i = 0; i < (int)selectedObjectIds.GetCount(); i++)
  {
    int objectId = selectedObjectIds.GetAt(i);
    stringStream << L"Object id: " << objectId << std::endl;
  }

  std::wstring resultOutput = stringStream.str();

  MessageBox(NULL,
    (LPCWSTR)resultOutput.c_str(),
    (LPCWSTR)L"Model Selection",
    MB_ICONINFORMATION | MB_OK);
}

SelectWallsButtonClickHandler::SelectWallsButtonClickHandler(Renga::IActionPtr pAction, Renga::IApplicationPtr pApplication)
  : Renga::ActionEventHandler(pAction),
    m_pApplication(pApplication)
{
}

void SelectWallsButtonClickHandler::OnTriggered()
{
  auto pProject = m_pApplication->GetProject();

  if (pProject == nullptr)
    return;

  auto pModel = pProject->GetModel();
  auto pModelObjects = pModel->GetObjects();

  CComSafeArray<int> modelObjectIds(pModelObjects->GetIds());

  std::vector<int> wallObjectIds;

  for (int i = 0; i < (int)modelObjectIds.GetCount(); i++)
  {
    int objectId = modelObjectIds.GetAt(i);

    auto pModelObject = pModelObjects->GetById(objectId);

    if (pModelObject == nullptr)
      continue;

    if (pModelObject->GetObjectType() != Renga::ObjectTypes::Wall)
      continue;

    wallObjectIds.push_back(objectId);
  }

  CComSafeArray<int> wallObjectIdsSafeArray((ULONG)wallObjectIds.size());

  for (int i = 0; i < wallObjectIds.size(); i++)
    wallObjectIdsSafeArray.SetAt(i, wallObjectIds[i]);

  auto pSelection = m_pApplication->GetSelection();
  pSelection->SetSelectedObjects(wallObjectIdsSafeArray);
}
